apps/meteor/app/push/server/fcm.ts

```diff
@@ -0,0 +1,159 @@
+import EJSON from 'ejson';
+import fetch from 'node-fetch';
+import type { RequestInit, Response } from 'node-fetch';
+
+import { settings } from '../../settings/server';
+import type { PendingPushNotification } from './definition';
+import { logger } from './logger';
+import type { FCMCredentials, NativeNotificationParameters } from './push';
+
+type FCMDataField = Record<string, any>;
+
+type FCMNotificationField = {
+	title: string;
+	body: string;
+	image?: string;
+};
+
+type FCMMessage = {
+	notification?: FCMNotificationField;
+	data?: FCMDataField;
+	token?: string;
+	to?: string;
+	android?: {
+		collapseKey?: string;
+		priority?: 'HIGH' | 'NORMAL';
+		ttl?: string;
+		restrictedPackageName?: string;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			analytics_label?: string;
+		};
+		direct_boot_ok?: boolean;
+	};
+	webpush?: {
+		headers?: FCMDataField;
+		data?: FCMDataField;
+		notification?: FCMNotificationField;
+		fcm_options?: {
+			link?: string;
+			analytics_label?: string;
+		};
+	};
+	fcm_options?: {
+		analytics_label?: string;
+	};
+};
+
+// https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
+type FCMError = {
+	error: {
+		code: number;
+		message: string;
+		status: string;
+	};
+};
+
+function fetchWithRetry(url: string, options: RequestInit, retries = 0): Promise<Response> {
+	const MAX_RETRIES = 5;
+	const ERRORS_TO_RETRY = [429, 500, 502, 503, 504];
+
+	try {
+		return fetch(url, options);
+	} catch (error) {
+		if (retries >= MAX_RETRIES) {
+			throw error;
+		}
+
+		const errorData = error as FCMError;
+
+		if (!errorData?.error?.code || !ERRORS_TO_RETRY.includes(errorData.error.code)) {
+			throw error;
+		}
+
+		return fetchWithRetry(url, options, retries + 1);
+	}
+}
+
+function getFCMMessagesFromPushData(userTokens: string[], notification: PendingPushNotification): { message: FCMMessage }[] {
+	// first we will get the `data` field from the notification
+	const data: FCMDataField = notification.payload ? { ejson: EJSON.stringify(notification.payload) } : {};
+
+	// Set image
+	if (notification.gcm?.image != null) {
+		data.image = notification.gcm?.image;
+	}
+
+	// Set extra details
+	if (notification.badge != null) {
+		data.msgcnt = notification.badge.toString();
+	}
+	if (notification.sound != null) {
+		data.soundname = notification.sound;
+	}
+	if (notification.notId != null) {
+		data.notId = notification.notId.toString();
+	}
+	if (notification.gcm?.style != null) {
+		data.style = notification.gcm?.style;
+	}
+
+	if (notification.contentAvailable != null) {
+		data['content-available'] = notification.contentAvailable.toString();
+	}
```

Comment:
I don't have any personal feud against `!=` over `!==`, but for consistency I'd ask to use `!==` or just simply `if (var) { ..`. At first glance I even typed out the comment `why checking for null, not being set I'd expect them to be undefined instead`, until noticed the coercion. I guess that is something. If the value could never be `null` explicitly, `!= null` is confusing when we use strict equals everywhere AFAIK. 

Depending on how serialized, you might even just `data.notId = notification.notId?.toString()` for instance. Depends on if serialization checks for truthness or existence (and also how receiver treats em, should be safe).

Reply:
I agree, I think I copied over this code from sendGcm
but that old pattern is probably a leftover from when it was written in js